cancel interaction only applies to interactions from same app

iCollin opened the pull request:
### Summary
Change CancelInteraction code so that an app may only cancel it's own interactions and not those of other apps.

#### Testing Plan
Cancel an interaction successfully
Try to cancel an interaction of another app (should be ignored)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -419,60 +419,59 @@ class UIController {
+                    this.onSystemContext("MAIN", rpc.params.appID)
+                    return true
+                } else if (rpc.params.functionID === 12 && app.alert.showAlert && !app.alert.isSubtle
                         && (rpc.params.cancelID === undefined || rpc.params.cancelID === app.alert.cancelID)) {
```
```suggestion
                    && (rpc.params.cancelID === undefined || rpc.params.cancelID === app.alert.cancelID)) {
```

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -419,60 +419,59 @@ class UIController {
+                    this.listener.send(RpcFactory.ScrollableMessageAbortedResponse(app.scrollableMessage.msgID));
+                    store.dispatch(closeScrollableMessage(app.alert.msgID, rpc.params.appID));
+                    const context = getNextSystemContext();
+                    this.onSystemContext(context, state.activeApp);
```
```suggestion
                    this.onSystemContext(context, rpc.params.appID);
```

iCollin replied:
I don't understand this, the app that is in full may have its context changed, not the App which sent CancelInteraction. It may have put an alert over another app. If sending app is in background then it's context isn't changed. Can you help me understand this suggestion?

jacobkeeler replied:
Hmmm, can't imagine a ScrollableMessage being shown from the background, but if we're accounting for that we'll probably need to change a few of these other cases as well.

iCollin replied:
The steal focus action always makes me consider the case of a background interaction.

I applied your suggestions, and double checked all the context changes sent from within this CancelInteraction handler are now the inverse of those applied when each interaction is first received.

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -490,60 +490,59 @@ class UIController {
+                    this.listener.send(RpcFactory.AlertAbortedResponse(app.alert.msgID))
+                    store.dispatch(closeAlert(app.alert.msgID, rpc.params.appID))
+                    const context = getNextSystemContext();
+                    this.onSystemContext(context, rpc.params.appID)
```
```suggestion
                    if (rpc.params.appID !== state.activeApp) {
                        this.onSystemContext("MAIN", rpc.params.appID)
                    }
                    this.onSystemContext(context, state.activeApp)
```

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -490,60 +490,59 @@ class UIController {
+                    this.listener.send(RpcFactory.SubtleAlertErrorResponse(app.alert.msgID, 5, 'subtle alert was cancelled'))
+                    store.dispatch(closeAlert(app.alert.msgID, rpc.params.appID))
+                    const context = getNextSystemContext();
+                    this.onSystemContext(context, rpc.params.appID)
```
```suggestion
                    if (rpc.params.appID !== state.activeApp) {
                        this.onSystemContext("MAIN", rpc.params.appID)
                    }
                    this.onSystemContext(context, state.activeApp)
```

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -490,60 +490,59 @@ class UIController {
+                    this.listener.send(RpcFactory.SliderAbortedResponse(app.slider.msgID))
+                    store.dispatch(closeSlider(app.alert.msgID, rpc.params.appID))
+                    const context = getNextSystemContext();
+                    this.onSystemContext(context, rpc.params.appID)
```
```suggestion
                    this.onSystemContext(context, state.activeApp)
```
If we're accounting for this message being sent in the background

jacobkeeler reviewed (changes requested):
Also will probably need to apply some of these changes to UI.ClosePopUp to match

jacobkeeler commented on src/js/Controllers/UIController.js:
```diff
@@ -490,60 +490,65 @@ class UIController {
                 
                 return { rpc: RpcFactory.UICancelInteractionIgnoredResponse(rpc) }
+            }
             case "ClosePopUp": {
```
We'll probably need to edit the `OnSystemContext` usages here too

iCollin replied:
Updated the calls from ClosePopUp in https://github.com/smartdevicelink/generic_hmi/pull/400/commits/d5533e893f784e9e88530620ac3df2c7ba27f089

iCollin commented:
In https://github.com/smartdevicelink/generic_hmi/pull/400/commits/cd3042a13f7d682d23b495bf8c7a07f81f44696a I aligned the context changes from the OnClose methods with these, ensuring we won't stick a background application in MENU context.

iCollin commented:
`onDefaultAction` and `onStealFocus ` were both meant to be called with the context parameter being the activeApp, but were sometimes called with the sending app as the context, so I have updated how those methods find the context in https://github.com/smartdevicelink/generic_hmi/pull/400/commits/05c893050cc89245daa0160791dfc770e324abb1 and https://github.com/smartdevicelink/generic_hmi/pull/400/commits/d91a612ee42bd681600636095efd7a6ac9619614.

I also modified Slider to send `HMI_OBSCURED` even if the sending app is in FULL in https://github.com/smartdevicelink/generic_hmi/pull/400/commits/05c893050cc89245daa0160791dfc770e324abb1.

jacobkeeler reviewed (changes requested):
```
        <element name="HMI_OBSCURED" internal_name="SYSCTXT_HMI_OBSCURED">
            <description>The app's display HMI is currently being obscured by either a system or other app's overlay.</description>
        </element>
```
Seems that this context is meant to be used when app is obscured by a dialog that is not its own

iCollin commented:
> Seems that this context is meant to be used when app is obscured by a dialog that is not its own

Interesting, thank you for pointing that out! I have reverted my change to HMI_OBSCURED for Slider and aligned ScrollableMessage to this new information as well.

PerformInteraction is still sending `HMI_OBSCURED` to the requesting app in both of our HMIs, so I did not mess with the context change here.
